app/src/main/java/com/brandon/forzenbook/viewmodels/LoginViewModel.kt

```diff
@@ -0,0 +1,291 @@
+
+    sealed interface LoginUiState {
+
+        data class Idle(
+            val email: String = "",
+            val isEmailError: Boolean = false,
+            val isGenericError: Boolean = false,
+            val isCodeSent: Boolean = false,
+        ) : LoginUiState
+
+        data class Loading(
+            val email: String = "",
+            val isCodeSent: Boolean = false,
+        ) : LoginUiState
+
+        // TODO remove this once we have another screen FA-83
+        object LoginWithCode : LoginUiState
+
+    }
+
+    sealed interface CreateAccountUiState {
+
+        data class Idle(
+            val email: String = "",
+            val firstName: String = "",
+            val lastName: String = "",
+            val location: String = "",
+            val dateOfBirth: String = "",
+            val isEmailError: Boolean = false,
+            val isLocationError: Boolean = false,
+            val isDateError: Boolean = false,
+            val isDuplicateUser: Boolean = false,
+        ) : CreateAccountUiState
+
+        data class Loading(
+            val email: String = "",
+            val firstName: String = "",
+            val lastName: String = "",
+            val location: String = "",
+            val dateOfBirth: String = "",
+        ) : CreateAccountUiState
+
+        object Loaded : CreateAccountUiState
+    }
+
+    fun getCode(email: String) {
+        viewModelScope.launch {
+            state.value = LoginUiState.Loading(isCodeSent = false, email = email)
+            if (!isValidEmail(email)) {
+                state.value = LoginUiState.Idle(isEmailError = true, email = email)
+                cancel()
+            } else {
+                if (getLoginCodeUseCase(email)) {
+                    state.value = LoginUiState.Idle(isCodeSent = true, email = email)
+                } else {
+                    state.value =
+                        LoginUiState.Idle(isGenericError = true, isCodeSent = false, email = email)
+                }
+                Log.e(VIEWMODEL_ERROR_TAG, "${state.value}")
+            }
+        }
+    }
+
+    fun loginClicked(email: String, code: String) {
+        viewModelScope.launch {
+            state.value = LoginUiState.Loading(isCodeSent = true, email = email)
+            if (!isValidEmail(email) || !isValidCode(code)) {
+                state.value = LoginUiState.Idle(isEmailError = true, email = email)
+                cancel()
+            } else {
+                if (loginUseCase(email, code)) {
+                    state.value = LoginUiState.LoginWithCode
+                } else {
+                    state.value = LoginUiState.Idle(isGenericError = true, email = email)
+                }
+            }
+        }
+        Log.e(VIEWMODEL_ERROR_TAG, "${state.value}")
+    }
+
+    fun createUser(
+        firstName: String,
+        lastName: String,
+        email: String,
+        date: String,
+        location: String
+    ) {
+        viewModelScope.launch {
```

Comment:
What thread do you intend to launch on? Should you specify IO specifically or?

Reply:
Now am specifying IO for these launches